From 3d861d63e13f8d187c06c5491f3da0fda597f223 Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Thu, 19 Dec 2024 19:54:34 +0800
Subject: [PATCH 3/4] mtk: mt76: mt7996: Change beacon protection fw cmd

Change beacon protection fw cmd from BSS_INFO to STA_REC_UPDATE
and fix memory leakage in mt7996_mcu_add_key

Signed-off-by: Allen Ye <allen.ye@mediatek.com>
(cherry picked from commit ac717989e4c26c4193c518842bcff878caf3e301)
---
 mt7996/main.c   | 14 ++++----
 mt7996/mcu.c    | 96 +++++++++++++++++--------------------------------
 mt7996/mcu.h    | 14 ++------
 mt7996/mt7996.h |  6 +++-
 4 files changed, 49 insertions(+), 81 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index 84c33598..7bec8100 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -682,6 +682,7 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 	int err = 0;
 	unsigned long add;
 	unsigned int link_id;
+	u8 pn[6] = {};
 
 	if (key->link_id >= 0) {
 		add = BIT(key->link_id);
@@ -728,6 +729,9 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 			if (key->keyidx == 6 || key->keyidx == 7) {
 				wcid_keyidx = &mlink->wcid.hw_key_idx2;
 				key->flags |= IEEE80211_KEY_FLAG_GENERATE_MMIE;
+				err = mt7996_mcu_get_pn(dev, mconf, mlink, pn);
+				if (err)
+					goto out;
 				break;
 			}
 			fallthrough;
@@ -738,6 +742,7 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 			return -EOPNOTSUPP;
 		}
 
+		/* Todo: remove me after fix set dtim period to fw */
 		if (cmd == SET_KEY && !sta && !mconf->mt76.cipher) {
 			mconf->mt76.cipher = mt76_connac_mcu_get_cipher(key->cipher);
 			mt7996_mcu_add_bss_info(mconf->phy, vif, conf, mconf, mlink, true);
@@ -753,12 +758,9 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 
 		mt76_wcid_key_setup(&dev->mt76, &mlink->wcid, key);
 
-		if (key->keyidx == 6 || key->keyidx == 7)
-			err = mt7996_mcu_bcn_prot_enable(dev, conf, mconf, mlink, key);
-		else
-			err = mt7996_mcu_add_key(&dev->mt76, mconf, key,
-						 MCU_WMWA_UNI_CMD(STA_REC_UPDATE),
-						 &mlink->wcid, cmd);
+		err = mt7996_mcu_add_key(&dev->mt76, mconf, key,
+					 MCU_WMWA_UNI_CMD(STA_REC_UPDATE),
+					 &mlink->wcid, cmd, pn);
 	}
 out:
 	mutex_unlock(&dev->mt76.mutex);
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index 677aba1d..89c3e326 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -3563,10 +3563,12 @@ int mt7996_mcu_teardown_mld_sta(struct mt7996_dev *dev,
 }
 
 static int
-mt7996_mcu_sta_key_tlv(struct mt76_wcid *wcid,
+mt7996_mcu_sta_key_tlv(struct mt76_dev *dev,
+		       struct mt76_wcid *wcid,
 		       struct sk_buff *skb,
 		       struct ieee80211_key_conf *key,
-		       enum set_key_cmd cmd)
+		       enum set_key_cmd cmd,
+		       u8 *pn)
 {
 	struct sta_rec_sec_uni *sec;
 	struct tlv *tlv;
@@ -3592,6 +3594,26 @@ mt7996_mcu_sta_key_tlv(struct mt76_wcid *wcid,
 		sec_key->key_len = key->keylen;
 		sec_key->need_resp = 0;
 		memcpy(sec_key->key, key->key, key->keylen);
+		/* Todo: move the BCN protection logic into mt76_connac_mcu_get_cipher() */
+		if (sec_key->key_id == 6 || sec_key->key_id == 7) {
+			switch (key->cipher) {
+			case WLAN_CIPHER_SUITE_AES_CMAC:
+				sec_key->cipher_id = MCU_CIPHER_BCN_PROT_CMAC_128;
+				break;
+			case WLAN_CIPHER_SUITE_BIP_GMAC_128:
+				sec_key->cipher_id = MCU_CIPHER_BCN_PROT_GMAC_128;
+				break;
+			case WLAN_CIPHER_SUITE_BIP_GMAC_256:
+				sec_key->cipher_id = MCU_CIPHER_BCN_PROT_GMAC_256;
+				break;
+			case WLAN_CIPHER_SUITE_BIP_CMAC_256:
+			default:
+				dev_err(dev->dev, "Unsupported BIGTK cipher\n");
+				return -EOPNOTSUPP;
+			}
+			sec_key->bcn_mode = BP_SW_MODE;
+			memcpy(sec_key->pn, pn, 6);
+		}
 
 		if (cipher == MCU_CIPHER_TKIP) {
 			/* Rx/Tx MIC keys are swapped */
@@ -3609,7 +3631,8 @@ mt7996_mcu_sta_key_tlv(struct mt76_wcid *wcid,
 
 int mt7996_mcu_add_key(struct mt76_dev *dev, struct mt7996_bss_conf *mconf,
 		       struct ieee80211_key_conf *key, int mcu_cmd,
-		       struct mt76_wcid *wcid, enum set_key_cmd cmd)
+		       struct mt76_wcid *wcid, enum set_key_cmd cmd,
+		       u8 *pn)
 {
 	struct sk_buff *skb;
 	int ret;
@@ -3619,17 +3642,18 @@ int mt7996_mcu_add_key(struct mt76_dev *dev, struct mt7996_bss_conf *mconf,
 	if (IS_ERR(skb))
 		return PTR_ERR(skb);
 
-	ret = mt7996_mcu_sta_key_tlv(wcid, skb, key, cmd);
-	if (ret)
+	ret = mt7996_mcu_sta_key_tlv(dev, wcid, skb, key, cmd, pn);
+	if (ret) {
+		dev_kfree_skb(skb);
 		return ret;
+	}
 
 	return mt76_mcu_skb_send_msg(dev, skb, mcu_cmd, true);
 }
 
-static int mt7996_mcu_get_pn(struct mt7996_dev *dev,
-			     struct ieee80211_bss_conf *conf,
-			     struct mt7996_bss_conf *mconf,
-			     struct mt7996_link_sta *mlink, u8 *pn)
+int mt7996_mcu_get_pn(struct mt7996_dev *dev,
+		      struct mt7996_bss_conf *mconf,
+		      struct mt7996_link_sta *mlink, u8 *pn)
 {
 #define TSC_TYPE_BIGTK_PN 2
 	struct sta_rec_pn_info *pn_info;
@@ -3661,60 +3685,6 @@ static int mt7996_mcu_get_pn(struct mt7996_dev *dev,
 	return 0;
 }
 
-int mt7996_mcu_bcn_prot_enable(struct mt7996_dev *dev,
-			       struct ieee80211_bss_conf *conf,
-			       struct mt7996_bss_conf *mconf,
-			       struct mt7996_link_sta *mlink,
-			       struct ieee80211_key_conf *key)
-{
-	struct mt7996_mcu_bcn_prot_tlv *bcn_prot;
-	struct sk_buff *skb;
-	struct tlv *tlv;
-	u8 pn[6] = {};
-	int len = sizeof(struct bss_req_hdr) +
-		  sizeof(struct mt7996_mcu_bcn_prot_tlv);
-	int ret;
-
-	skb = __mt7996_mcu_alloc_bss_req(&dev->mt76, &mconf->mt76, len);
-	if (IS_ERR(skb))
-		return PTR_ERR(skb);
-
-	tlv = mt76_connac_mcu_add_tlv(skb, UNI_BSS_INFO_BCN_PROT, sizeof(*bcn_prot));
-
-	bcn_prot = (struct mt7996_mcu_bcn_prot_tlv *)tlv;
-
-	ret = mt7996_mcu_get_pn(dev, conf, mconf, mlink, pn);
-	if (ret) {
-		dev_kfree_skb(skb);
-		return ret;
-	}
-
-	switch (key->cipher) {
-	case WLAN_CIPHER_SUITE_AES_CMAC:
-		bcn_prot->cipher_id = MCU_CIPHER_BCN_PROT_CMAC_128;
-		break;
-	case WLAN_CIPHER_SUITE_BIP_GMAC_128:
-		bcn_prot->cipher_id = MCU_CIPHER_BCN_PROT_GMAC_128;
-		break;
-	case WLAN_CIPHER_SUITE_BIP_GMAC_256:
-		bcn_prot->cipher_id = MCU_CIPHER_BCN_PROT_GMAC_256;
-		break;
-	case WLAN_CIPHER_SUITE_BIP_CMAC_256:
-	default:
-		dev_err(dev->mt76.dev, "Not supported Bigtk Cipher\n");
-		dev_kfree_skb(skb);
-		return -EOPNOTSUPP;
-	}
-
-	pn[0]++;
-	memcpy(bcn_prot->pn, pn, 6);
-	bcn_prot->enable = BP_SW_MODE;
-	memcpy(bcn_prot->key, key->key, WLAN_MAX_KEY_LEN);
-	bcn_prot->key_id = key->keyidx;
-
-	return mt76_mcu_skb_send_msg(&dev->mt76, skb,
-				     MCU_WMWA_UNI_CMD(BSS_INFO_UPDATE), true);
-}
 int mt7996_mcu_add_dev_info(struct mt7996_phy *phy,
 			    struct ieee80211_bss_conf *conf,
 			    struct mt7996_bss_conf *mconf, bool enable)
diff --git a/mt7996/mcu.h b/mt7996/mcu.h
index 68dbcb43..9441a812 100644
--- a/mt7996/mcu.h
+++ b/mt7996/mcu.h
@@ -467,17 +467,6 @@ enum {
 	BP_HW_MODE,
 };
 
-struct mt7996_mcu_bcn_prot_tlv {
-	__le16 tag;
-	__le16 len;
-	u8 pn[6];
-	u8 enable;
-	u8 cipher_id;
-	u8 key[WLAN_MAX_KEY_LEN];
-	u8 key_id;
-	u8 __rsv[3];
-} __packed;
-
 struct bss_ra_tlv {
 	__le16 tag;
 	__le16 len;
@@ -711,6 +700,9 @@ struct sec_key_uni {
 	u8 key_len;
 	u8 need_resp;
 	u8 key[32];
+	u8 pn[6];
+	u8 bcn_mode;
+	u8 _rsv;
 } __packed;
 
 struct sta_rec_sec_uni {
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index 76215f68..e664f584 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -1463,9 +1463,13 @@ int mt7996_init_dev_debugfs(struct mt7996_phy *phy);
 int mt7996_init_band_debugfs(struct mt7996_phy *phy);
 void mt7996_debugfs_rx_fw_monitor(struct mt7996_dev *dev, const void *data, int len);
 bool mt7996_debugfs_rx_log(struct mt7996_dev *dev, const void *data, int len);
+int mt7996_mcu_get_pn(struct mt7996_dev *dev,
+		      struct mt7996_bss_conf *mconf,
+		      struct mt7996_link_sta *mlink, u8 *pn);
 int mt7996_mcu_add_key(struct mt76_dev *dev, struct mt7996_bss_conf *mconf,
 		       struct ieee80211_key_conf *key, int mcu_cmd,
-		       struct mt76_wcid *wcid, enum set_key_cmd cmd);
+		       struct mt76_wcid *wcid, enum set_key_cmd cmd,
+		       u8 *pn);
 int mt7996_mcu_bcn_prot_enable(struct mt7996_dev *dev,
 			       struct ieee80211_bss_conf *conf,
 			       struct mt7996_bss_conf *mconf,
-- 
2.45.2

